todo-list-backend/tests/integration/graphql-api.spec.ts

```diff
@@ -0,0 +1,526 @@
+import {buildApolloServer} from "../../src";
+import {InMemoryStorage, IStorage} from "../../src/data";
+import {User, ToDo} from "../../src/types";
+import {ApolloServerTestClient, createTestClient} from "apollo-server-testing";
+import {gql} from "apollo-server";
+import {sign} from "jsonwebtoken";
+
+const READ_TODOS = gql`query { readTodos { id name done assignee { name }}} `;
+
+const LOGIN = gql`
+    mutation Login($name: String!, $password: String!) {
+        login(name: $name, password: $password)
+    }
+`;
+
+const CREATE_TODO = gql`
+    mutation CreateTodo($name: String!) {
+        createTodo(name: $name) {
+            id name done assignee { name }
+        }
+    }
+`;
+
+const UPDATE_TODO_NAME = gql`
+    mutation UpdateTodoName($id: Int!, $name: String!) {
+        updateTodoName(id: $id, name: $name) {
+            id name done assignee { name }
+        }
+    }
+`;
+
+const UPDATE_TODO_DONE = gql`
+    mutation UpdateTodoDone($id: Int!, $done: Boolean!) {
+        updateTodoDone(id: $id, done: $done)  {
+            id name done assignee { name }
+        }
+    }
+`;
+
+const DELETE_TODO = gql`
+    mutation DeleteTodo($id: Int!) {
+        deleteTodo(id: $id)  {
+            id name done assignee { name }
+        }
+    }
+`;
+
+const CREATE_USER = gql`
+    mutation CreateUser($name: String!, $password: String!) {
+        createUser(name: $name, password: $password) {
+            name
+        }
+    }
+`;
+
+function buildTwoUserInMemoryStorage(addTodos: boolean = false): IStorage {
+    let initialUsers = [
+        new User("ralph", "ralph"),
+        new User("nora", "nora")
+    ];
+
+    if (!addTodos) {
+        return new InMemoryStorage(initialUsers);
+    }
+
+    let initialTodos = [
+        new ToDo("first todo", initialUsers[1]),
+        new ToDo("second todo", initialUsers[0]),
+    ];
+
+    return new InMemoryStorage(initialUsers, initialTodos);
+
+}
+
+describe("query", () => {
+    const authSecret: string = "This is a perfect secret for testing!";
+    let user: User;
+    let storage: IStorage;
+    let query: any;
+    let mutate: any;
+
+
+    describe("given a legit authorization token", () => {
+        beforeEach(async () => {
+            storage = buildTwoUserInMemoryStorage(true);
+
+            // workaround for the shortcomings of the apollo-server-testing
+            // package as it doesn't generate a real request and therefor a
+            // "real" context will never be generated
+            user = storage.readUser("ralph");
+            const token = sign({username: user.name}, authSecret, {expiresIn: "1 day"});
+            const context = {token: token, user: user};
+
+            const testServer = buildApolloServer(storage, authSecret, false, context);
+            const testClient = createTestClient(testServer);
```

Comment:
Just one thing. If you happen to use middlewares and you re-create your apollo server like that in a `beforeEach` you likely run into https://github.com/prisma-labs/graphql-middleware/issues/63#issuecomment-520786385.

Luckily, it still works if you create the `testServer` and `testClient` in a `beforeAll` (that will only apply the middlewares once) and you can still change the data in your database for each test case. Your tests will even be slightly faster because they initiate only one `testServer` per test suite.